src/backend/distributed/planner/distributed_planner.c

```diff
@@ -126,12 +129,24 @@ distributed_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 		 * set, which doesn't break our goals, but, prevents us keeping an extra copy
 		 * of the query tree. Note that we copy the query tree once we're sure it's a
 		 * distributed query.
+		 *
+		 * Fast path router queries don't need RTE identities or partitioning adjustments
+		 * because they're only relevant when planning relies on restrict information.
 		 */
-		AssignRTEIdentities(parse);
-		originalQuery = copyObject(parse);
+		if (!fastPathRouterQuery)
+		{
+			AssignRTEIdentities(parse);
+			originalQuery = copyObject(parse);
 
-		setPartitionedTablesInherited = false;
-		AdjustPartitioningForDistributedPlanning(parse, setPartitionedTablesInherited);
+			setPartitionedTablesInherited = false;
+			AdjustPartitioningForDistributedPlanning(parse,
+													 setPartitionedTablesInherited);
```

Comment:
I'm slightly concerned that we might introduce subtle bugs in the future because the query tree might be in different states in the router planner.

Worth noting that these functions can be a lot faster:
https://github.com/citusdata/citus/commit/83f81e1fe215c3e79a49927d4d7234e4866f229d

Reply:
> Worth noting that these functions can be a lot faster:
83f81e1

Agreed, I think if we decide to get `faster_all_the_things` branch, merging this would not be as affective mentioned in the PR description.